sticker-utils/src/bin/sticker-train.rs

```diff
@@ -22,9 +22,71 @@ static LR_SCALE: &str = "LR_SCALE";
 static LR_PATIENCE: &str = "LR_PATIENCE";
 static CONTINUE: &str = "CONTINUE";
 static PATIENCE: &str = "PATIENCE";
+static SAVE_BATCH: &str = "SAVE_BATCH";
 static TRAIN_DATA: &str = "TRAIN_DATA";
 static VALIDATION_DATA: &str = "VALIDATION_DATA";
 
+#[derive(Copy, Clone, Eq, PartialEq)]
+pub enum CompletedUnit {
+    /// A batch is completed.
+    Batch,
+
+    /// An epoch is completed.
+    Epoch,
+}
+
+#[derive(Copy, Clone, Eq, PartialEq)]
+pub enum SaveSchedule {
+    /// Save after every epoch.
+    Epoch,
+
+    /// Save after every N batches.
+    Batches(usize),
+}
+
+/// Scheduler that saves at points dictated by the schedule.
+pub struct SaveScheduler {
+    epoch: usize,
+    batch: usize,
+    schedule: SaveSchedule,
+}
+
+impl SaveScheduler {
+    /// Save the model paramters when a save point has been reached.
+    fn save(&mut self, trainer: &TaggerTrainer, completed: CompletedUnit) -> Fallible<()> {
+        match completed {
+            CompletedUnit::Epoch => {
+                if self.schedule == SaveSchedule::Epoch {
+                    trainer.save(format!("epoch-{}", self.epoch))?;
+                }
+
+                self.epoch += 1;
+            }
+            CompletedUnit::Batch => {
+                if let SaveSchedule::Batches(batches) = self.schedule {
+                    if (self.batch + 1) % batches == 0 {
+                        trainer.save(format!("batch-{}", self.batch))?
+                    }
+                }
+
+                self.batch += 1;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl From<SaveSchedule> for SaveScheduler {
+    fn from(schedule: SaveSchedule) -> Self {
+        SaveScheduler {
+            batch: 0,
+            epoch: 0,
+            schedule,
+        }
+    }
+}
+
```

Comment:
Maybe reorder the methods and structs in this file such that main is either at the top or the bottom of the file?

Reply:
Good idea!

Reply:
I guess the ordering of `main` in the middle was already started by previous commits. So I'll do a separate clean up PR.